FR #1094: Add Dataset Completeness attribute for custom columns (#1800)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/glossary.md

A financial incentive or allowance granted by a service provider unrelated to other past/current/future charges.

<a name="glossary:custom-column"><b>Custom Column</b></a>

> A column in a [*FOCUS dataset*](#glossary:FOCUS-dataset) that is not defined by FOCUS. Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns.

Reviewer: I would really like the first sentence of this definition to scan the same as the FOCUS column definition:

> A column defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset).

Therefore:

```suggestion
A column not defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns.
```

It's either that, or keep the pattern here, and change the FOCUS column definition to follow _this_ flow.

> A column not defined by FOCUS and included a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns.

Reviewer: ```suggestion
A column not defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns.
```

Author: 🤖 [AI][Claude] ✅ **Implemented**

Fixed the typo ("included a" → "included in a"). Kept the rest of the entry as-is.

A financial incentive or allowance granted by a service provider unrelated to other past/current/future charges.

<a name="glossary:custom-column"><b>Custom Column</b></a>

> A column not defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns.

Reviewer: This definition is good, but since we are locking down the language, you might want to add a cross-reference directly to the new attribute so readers know exactly which custom columns are required.  Consider appending this reference to the end:

```suggestion
A column not defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns. See [Dataset Completeness](#attributes.datasetcompleteness) for inclusion requirements.
```

Author: 🤖 [AI][Claude] ✅ **Implemented**

Added Dataset Completeness cross-reference. The Custom Column glossary entry now reads:

> A column not defined by FOCUS and included in a *FOCUS dataset*. Custom columns are prefixed with `x_` and provide additional context from *native datasets* beyond what is captured in FOCUS columns. See Dataset Completeness for inclusion requirements.

Reviewer: Looks good!

An open-source specification that defines requirements for billing data.

<a name="glossary:FOCUS-column"><b>FOCUS Column</b></a>

A column defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). See the Columns section of each dataset for definitions.

<a name="glossary:FOCUS-dataset"><b>FOCUS Dataset</b></a>

A structured collection of columns that conforms to the BCP14 criteria established by FOCUS. All columns included must be defined in the FOCUS Columns section of the specification.

> In addition to these standardized columns, [data generators](#metadata.datagenerator) may include [*custom columns*](#glossary:custom-column) where additional context is needed beyond what is captured in the defined [*FOCUS columns*](#glossary:FOCUS-column). If custom columns introduce record-splitting (i.e., a single original charge results in multiple rows), the data generator is responsible for ensuring that all cost and quantity metrics still meet the aggregation and consistency rules required by the specification.

Reviewer: This needs a tweak because the whole point of this PR is to require the inclusion of custom columns -- it's no longer an optional endeavor.  That said, I don't think we should say "must" or "should" here -- we merely strike the word "may".

```suggestion
In addition to these standardized columns, [data generators](#metadata.datagenerator) include [*custom columns*](#glossary:custom-column) to capture information from [*native datasets*](#glossary:native-dataset) that is not represented by [*FOCUS columns*](#glossary:FOCUS-column). If custom columns introduce record-splitting (i.e., a single original charge results in multiple rows), the data generator is responsible for ensuring that all cost and quantity metrics still meet the aggregation and consistency rules required by the specification.
```

Author: 🤖 [AI][Claude] ✅ **Implemented**

Struck "may" and updated the wording. The FOCUS Dataset glossary entry now reads:

> In addition to these standardized columns, data generators include *custom columns* to capture information from *native datasets* that is not represented by *FOCUS columns*.

Reviewer: Looks good!

A government-issued currency (e.g., US dollars, Euros).

<a name="glossary:native-dataset"><b>Native Dataset</b></a>

> A dataset provided by a [*data generator*](#metadata.datagenerator) in a format other than FOCUS. For [*service providers*](#glossary:service provider), this typically refers to their proprietary billing export (e.g., AWS Cost and Usage Report, Azure Cost Details, GCP BigQuery Billing Export). For FinOps tool vendors, this refers to any non-FOCUS dataset they offer to [*practitioners*](#glossary:practitioner).

Reviewer 2: **Notes:**

- we should add service-provider and other participating entities glossary entries (discussed in TF-1 as well)
- just realized that in specification/overview.md, line 5 we have a similar broken link ("*...infrastructure and service [*service providers*](#glossary:service provider)...*")
- consider adding **FinOps Tool Vendors** or **FinOps Vendors** glossary entry

```suggestion
A dataset provided by a [*data generator*](#metadata.datagenerator) in a format other than FOCUS. For [*service providers*](#glossary:service-provider), this typically refers to their proprietary billing export (e.g., AWS Cost and Usage Report, Azure Cost Details, GCP BigQuery Billing Export). For FinOps tool vendors, this refers to any non-FOCUS dataset they offer to [*practitioners*](#glossary:practitioner).
```

Author: 🤖 [AI][Claude] 🤔 **Needs discussion**

The suggested expansion of the native dataset glossary definition looks reasonable — it already mentions FinOps tool vendors in the current text.

However, the broader glossary work (adding service-provider, FinOps tool vendor entries, fixing the broken link in `overview.md`) seems out of scope for this PR. These would benefit from a dedicated issue or inclusion in a broader glossary cleanup effort.

Should we track these as a separate issue? The `overview.md` broken link in particular seems like it predates this PR.

Author: 🤖 [AI][Claude] ✅ **Implemented**

- Fixed the `#glossary:service provider` → `#glossary:service-provider` anchor in the native dataset glossary entry
- Broadened the definition to not be CostAndUsage-specific: "proprietary data exports (e.g., billing exports, contract details, or other FinOps-related datasets)" instead of listing specific provider billing exports
- The FinOps tool vendor mention was already present in the definition and remains unchanged
- Fixed the broken `overview.md` link (commit 74fe5eee)

A government-issued currency (e.g., US dollars, Euros).

<a name="glossary:native-dataset"><b>Native Dataset</b></a>

> A dataset provided by a [*data generator*](#metadata.datagenerator) in a format other than FOCUS. For [*service providers*](#glossary:service-provider), this typically refers to their proprietary data exports (e.g., billing exports, contract details, or other FinOps-related datasets). For FinOps tool vendors, this refers to any non-FOCUS dataset they offer to [*practitioners*](#glossary:practitioner).

Reviewer: I would like the beginning of the `Native Dataset` entry to read in the same format as the beginning of the `FOCUS Dataset` entry.  Can you take a stab at revising both those entries to read similarly -- and to link to each other, given that they don't alpha-sort next to each other?

Author: 🤖 [AI][Claude] ✅ **Implemented**

Revised both entries to have parallel structure. Added cross-links ("See also: Native Dataset" on FOCUS Dataset, "See also: FOCUS Dataset" on Native Dataset). Also added glossary entries for FOCUS Column and ensured the existing Custom Column entry is consistent.
